### PR TITLE
(PTFE-2504) develop branch with three digits

### PR DESCRIPTION
**Pull request by matthiasL-scality**
# PR: Support for development/x.y.z Branches (Three-Digit Development Branches)

## Summary
This pull request introduces full support for three-digit development branches (e.g., `development/x.y.z`) in the GitWaterFlow cascade logic. The changes ensure that the system correctly handles, sorts, and calculates next versions for both two-digit (`development/x.y`) and three-digit (`development/x.y.z`) development branches.

## Key Changes
- **Cascade Logic:**
  - The cascade now properly includes and orders three-digit development branches alongside their two-digit parents.
  - When the destination is a three-digit branch, its parent two-digit branch is always included in the cascade, and the order is: three-digit branch first, then its parent, then the rest.
  - The default sorting ensures that for the same major/minor, the two-digit branch always comes before any three-digit branch.

- **Next Micro Version Calculation:**
  - For two-digit branches, the next micro version is now always the first available micro not already claimed by a three-digit branch, regardless of whether that branch is ignored in the cascade.
  - The logic skips all micro versions for which a `development/x.y.z` branch exists, ensuring no conflicts or duplicate versions.
  - Each target version is only added once to the list, preventing duplicates when both a two-digit and a three-digit branch would propose the same next version.

- **Tests:**
  - The test suite has been updated to cover all edge cases, including mixed two-digit and three-digit development branches, correct cascade ordering, and version calculation.
  - Tests now ensure that the cascade and version logic behave as expected for all combinations of present and ignored branches.

## Impact
- No breaking changes for existing two-digit or major-only development branch workflows.
- Three-digit development branches are now fully supported and integrated into the cascade and versioning logic.
- Improved reliability and predictability for release and merge automation.

---

### CODE REVIEW

File: bert_e/tests/test_bert_e.py [hunk 1]
```diff
@@ -190,6 +190,66 @@ def dynamic_filtering(request):
...
+        from bert_e.workflow.gitwaterflow.branches import compare_branches
+        branch1 = ((4, 3, 2),)
+        branch2 = ((4, 3),)
+        self.assertEqual(compare_branches(branch1, branch2), 2)
+
+    def test_compare_branches_major_minor_vs_major_minor_micro(self):
+        from bert_e.workflow.gitwaterflow.branches import compare_branches
+        branch1 = ((4, 3),)
+        branch2 = ((4, 3, 2),)
+        self.assertEqual(compare_branches(branch1, branch2), -2)
```
Comment (tcarmet) [inline bert_e/tests/test_bert_e.py:221]: Add those tests inside this file: https://github.com/scality/bert-e/blob/main/bert_e/tests/unit/test_sorted.py

and no need to keep importing `compare_branches` (already done in test_sorted at the top of the file)

File: bert_e/tests/test_bert_e.py [hunk 2]
```diff
@@ -190,6 +190,66 @@ def dynamic_filtering(request):
...
+        from bert_e.exceptions import MasterQueueMissing
+        exc = MasterQueueMissing((10, 0, 5))
+        self.assertIn('q/10.0.5 is missing', str(exc))
+
+    def test_master_queue_missing_fallback(self):
+        # Covers fallback case
+        from bert_e.exceptions import MasterQueueMissing
+        exc = MasterQueueMissing([])
+        self.assertIn('q/[] is missing', str(exc))
+
```
Comment (tcarmet) [inline bert_e/tests/test_bert_e.py:252]: Similar story here, you can create a new file called: `bert_e/tests/unit/test_master_queue_missing.py` import `MasterQueueMissing` on top of the file and move those tests there. 

The overall idea is to try to avoid adding unit test into this file that to reduce its size cause it's huge already :D

File: bert_e/tests/test_bert_e.py [hunk 3]
```diff
@@ -450,10 +510,37 @@ def test_branch_cascade_target_hotfix(self):
         fixver = ['6.6.6.3']
         self.finalize_cascade(branches, tags, destination, fixver)
 
-    def test_branch_cascade_invalid_dev_branch(self):
+    def test_branch_cascade_target_three_digit_dev(self):
+        """Test cascade targeting three-digit development branch"""
         destination = 'development/4.3.17'
         branches = OrderedDict({
-            1: {'name': 'development/4.3.17', 'ignore': False}
+            1: {'name': 'development/4.3', 'ignore': True},
```
Comment (tcarmet) [inline bert_e/tests/test_bert_e.py:517]: Shouldn't it be the following? 

```suggestion
            1: {'name': 'development/4.3', 'ignore': False},
```

File: bert_e/tests/test_bert_e.py [hunk 4]
```diff
@@ -450,10 +510,37 @@ def test_branch_cascade_target_hotfix(self):
...
+    def test_branch_cascade_mixed_versions(self):
+        """Test cascade with mix of 2-digit and 3-digit development branches"""
+        destination = 'development/5.1'
+        branches = OrderedDict({
+            1: {'name': 'development/4.3', 'ignore': True},
+            2: {'name': 'development/4.3.17', 'ignore': True},
+            3: {'name': 'development/5.1', 'ignore': False},
+            4: {'name': 'development/5.1.8', 'ignore': False},
+            5: {'name': 'development/10.0', 'ignore': False}
+        })
```
Comment (tcarmet) [inline bert_e/tests/test_bert_e.py:535]: Similar story, I expect `5.1.8` to be ignored if `development/5.1` is targetted:

```suggestion
        branches = OrderedDict({
            1: {'name': 'development/4.3.17', 'ignore': True},
            2: {'name': 'development/4.3', 'ignore': True},
            3: {'name': 'development/5.1.8', 'ignore': True},
            4: {'name': 'development/5.1', 'ignore': False},
            5: {'name': 'development/10.0', 'ignore': False}
        })
```

**Review by tcarmet — CHANGES_REQUESTED** [review]
Glad you have some tests now, this gives me a better view of what's up.

Something is up with how you are comparing branches I think. Similar to how `development/x` are > than `development/x.y` the `development/x.y` should be > than the `development/x.y.z`.
I say this because I see `development/x.y` getting ignored when the destination is `development/x.y.z` where I don't think it should be.

What I also think it's missing in terms of tests:

- Are we capable of handling more than one `development/x.y.z` (I think we are, but need to see it in action). For example `development/4.1.17` with `development/4.1.18` and `development/4.1`.
- When the tag `4.1.18` exists and we are working with `development/4.1.18` what are we asking in fix version?
- (This one might deserve its own PR and we can try to do it together to minimize changes needed) we should be adding a three digit branch on the base git repo init scenario. Now a whole bunch of tests will break because of it, but it should allow us to identify if we broke something.

  - Kind of related to the above so will implicitly fix itself: Haven't seen any dedicated functional tests with `dev/x.y.z` branches.

- Since you kept the hotfix branches, if `hotfix/x.y.z` branches PR + `development/x.y.z` are into the mix together are we behaving correctly? Can we queue them together? (having a pr of each in the queue)

File: bert_e/tests/test_bert_e.py (hunk 3 above)
Comment (charlesprost) [reply bert_e/tests/test_bert_e.py:517]: Indeed, dev/4.3.17 goes into dev/4.3

File: bert_e/tests/test_bert_e.py (hunk 4 above)
Comment (charlesprost) [reply bert_e/tests/test_bert_e.py:535]: Agree

File: bert_e/workflow/gitwaterflow/branches.py [hunk 5]
```diff
@@ -30,23 +30,57 @@
...
+
+    if is_major_only_1 and not is_major_only_2:
+        # version1 is major-only, version2 is major.minor
+        # major-only comes after
+        return 1
+    elif not is_major_only_1 and is_major_only_2:
+        # version1 is major.minor, version2 is major-only
+        # major.minor comes before
+        return -1
+    elif is_major_only_1 and is_major_only_2:
```
Comment (charlesprost) [inline bert_e/workflow/gitwaterflow/branches.py:64]: Should be a bug to have the same (4,None) for both no?

File: bert_e/workflow/gitwaterflow/branches.py [hunk 6]
```diff
@@ -156,38 +190,66 @@ def version_t(self):
...
+        if not hasattr(self, 'micro') or self.micro is None:
+            self.micro = None  # Keep as None for x.y format branches
+
+        # Update version property to include micro when present
+        if self.micro is not None:
+            self.version = f"{self.major}.{self.minor}.{self.micro}"
+        elif self.minor is not None:
+            self.version = f"{self.major}.{self.minor}"
+        else:
+            self.version = str(self.major)
```
Comment (charlesprost) [inline bert_e/workflow/gitwaterflow/branches.py:216]: Cast the same way

File: bert_e/workflow/gitwaterflow/branches.py [hunk 7]
```diff
@@ -156,38 +190,66 @@ def version_t(self):
...
-        return self.minor < other.minor
+
+        # Same major version - apply cascade rules
+
+        if self.minor is not None and other.minor is None:
+            return True   # self (has minor) < other (no minor)
+
+        # Both have minor versions
+        if self.minor != other.minor:
+            return self.minor < other.minor
```
Comment (charlesprost) [inline bert_e/workflow/gitwaterflow/branches.py:239]: What is self.minor is None? You have lost this case compared to before

**Review by charlesprost — CHANGES_REQUESTED** [review]

**Review by tcarmet — COMMENTED** [review]
Looking better, still need to see the test cases I mentioned, I added one as a bonus on my review.

For the more functional test setup, we can set one up together in a session if you need a hand on that, over a 30 min session we can get this sorted easily. Just let me know :)

File: bert_e/tests/test_bert_e.py [hunk 8]
```diff
@@ -450,10 +450,82 @@ def test_branch_cascade_target_hotfix(self):
...
+            1: {'name': 'development/4.3.17', 'ignore': True},
+            2: {'name': 'development/4.3', 'ignore': True},
+            3: {'name': 'development/5.1.8', 'ignore': True},
+            4: {'name': 'development/5.1', 'ignore': False},
+            5: {'name': 'development/10.0', 'ignore': False}
+        })
+        tags = ['4.3.16', '4.3.18', '5.1.3', '5.1.7']
+        fixver = ['5.1.9', '10.0.0']
+        self.finalize_cascade(branches, tags, destination, fixver)
+
```
[the remark this answers is not in the document]
Comment (tcarmet) [reply bert_e/tests/test_bert_e.py:524]: we'll have to check but bert-e should already be able to interpret the `4.3.17.0` correctly  (if you look at how the ring is tagged you'll only see 4 digits) and we have to return what I said in the fixversion, or actually check what is expected in this case and set expectations right.

[the remark this answers is not in the document]
Comment (tcarmet) [reply bert_e/tests/test_bert_e.py:558]: Ok thanks for reviewing that. So that is completely fine in terms of behavior, I was under the impression though that you wanted to make this a long living branch. (in a way that they can eventually replace hotfix branches/make them obsolete).

So in that case my test case is not supposed to pass as is, and I'll be suggesting others that point us towards the expected behavior. 

You most likely already know this @charlesprost but just in case it wasn't clear, this means when we go GA on this tag, we have to quickly delete the branch otherwise we block all other PRs from merging (similar situation we had when we had the stab).

[the remark this answers is not in the document]
Comment (charlesprost) [reply bert_e/tests/test_bert_e.py:558]: @tcarmet yeah, it was a possibility we could have to live with long living branches, but indeed that was not our intention, therefore we now raise an error in this case.

And you are correct, it means as soon as we are GA, Bert-E will get stuck as we did before with the stab branch, asking for the dev/x.y.z to be deleted before it can proceed with any other PR. That's fine by me as I prefer this hard stop rather than allowing a ticket with 9.5.0.1 as fixver to go into dev/9.5.0 post-GA with the cascade and everything :sweat:

**Review by tcarmet — APPROVED** [review]
I think those two topics need to be solved before merging (second one could arguably be optional but I rather be safe):

- I understand why you removed the functional test we setup, but now we have no functional test that tries to directly target a dev/x.y.z branch as destination. It should work fine, but we **have** to explicitly test it.
- I also think we implicitly test it, but I do want to see explicitly if we are capable of handling both a hotfix PR and a dev/x.y.z PR in the queue together.

I understand you want to merge this soon, therefore approving but take those into account please.
